docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+tokens are returned to the owner after stake delegation has finished.
+Owner’s participation is not required in the day-to-day operations on the
+stake, so cold storage can be accommodated to the maximum extent.
+
+Operator:: An address of a party authorized to operate in the network on behalf
+of a given owner. The operator handles the everyday operations on the delegated
+stake without actually owning the staked tokens. An operator can not simply
+transfer away delegated tokens, however, it should be noted that the operator’s
+misbehavior may result in slashing tokens and thus the entire staked amount is
+indeed at stake.
+
+Beneficiary:: An address where the rewards for participation are sent, earned by
+the operator, on behalf of the owner. A beneficiary doesn’t sign or publish any
+protocol-relevant transactions, but any currency or tokens earned by the
+operator will be transferred to the beneficiary.
+
+Authorizer:: An address appointed by an owner to authorize applications on
+behalf of them. An application must be approved by the authorizer before the
+operator is eligible to participate.
+
+Delegated stake:: An owner’s staked tokens, delegated to the operator by the
+owner. Delegation enables KEEP owners to have their wallets offline and their
+stake operated by operators on their behalf.
+
+== Specification
+
+=== Functionality
+
+==== Delegating a stake
+
+Tokens are delegated by the owner. During the delegation, the owner needs to
+appoint an operator, beneficiary, and authorizer. The owner may decide to
+delegate just a portion of their tokens. The owner may delegate multiple times
+to different operators. For simplicity, the operator address can not be reused
+between delegations even if the delegation has been recovered. The only
+exception is when the delegation has been canceled. To protect against griefing
+by frontrunning stake delegation transactions and reserving operator addresses,
+staking contract should require a governable minimum stake amount for each
+delegation.
+
+Staking contract binds the owner with the operator, beneficiary, and authorizer
+for the given stake delegation and records the stake delegation amount.
+Delegation is set to active after 2 hours, once the cancellation window passes.
+
+Existing KEEP and NU stake owners also need to perform a delegation so that T
+staking contract can record their operator, beneficiary, and authorizer. Since
+these values are read-only, they should be copied from KEEP staking contract to
+T staking contract and NU stakers should set them during the delegation. For
+existing stakes, the entire stake amount is registered in T staking contract.
+
+There are 2 hours to cancel a stake without suffering an undelegation period.
+This functionality is needed e.g. in case of a typo in the operator address.
+Undelegation can not be canceled in the same transaction in which it was
+created. Once the delegation has been canceled, the operator address can be used
+again.
```

Comment:
We should note that stake that can be cancelled can not be authorised for any application, to prevent risk-and-obligation-free drive-by reward sniping where someone joins a sortition pool for just long enough to grab a reward and then cancels the stake and removes themselves from the pool without suffering an undelegation period.

Reply:
👌🏻 well said